openfe/setup/atommapping.py

```diff
@@ -19,14 +24,99 @@ class AtomMapping:
       If this atom has no corresponding atom, None is returned.
 
     """
+
     mol1: Molecule
     mol2: Molecule
-    mol1_to_mol2: Dict[int, int]
+    mol1_to_mol2: AtomMap
 
     def __hash__(self):
-        return hash((hash(self.mol1), hash(self.mol2),
-                     tuple(self.mol1_to_mol2.items())))
+        return hash(
+            (hash(self.mol1), hash(self.mol2), tuple(self.mol1_to_mol2.items()))
+        )
 
     @classmethod
     def from_perses(cls, perses_mapping):
         raise NotImplementedError()
+
+    def _match_elements(self, idx: int):
+        elem_mol1 = self.mol1.GetAtomWithIdx(idx).GetAtomicNum()
+        elem_mol2 = self.mol2.GetAtomWithIdx(self.mol1_to_mol2[idx]).GetAtomicNum()
+        if elem_mol1 == elem_mol2:
+            return True
+        else:
+            return False
+
+    def _get_unique_bonds_and_atoms(self, mapping: AtomMap):
+
+        uniques = {
+            "atoms": [],  # atoms which fully don't exist in mol2
+            "elements": [],  # atoms which exist but change elements in mol2
+            "bonds": [],  # bonds involving either unique atoms or elements
+        }
+
+        for at in self.mol1.GetAtoms():
+            idx = at.GetIdx()
+            if idx not in self.mol1_to_mol2.keys():
+                uniques["atoms"].append(idx)
+            elif not self._match_elements(idx):
+                uniques["elements"].append(idx)
+
+        for bond in self.mol1.GetBonds():
+            bond_at_idxs = [bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()]
+            for at in chain(uniques["atoms"], uniques["elements"]):
+                if at in bond_at_idxs:
+                    bond_idx = bond.GetIdx()
+                    # avoid double inclusion
+                    if bond_idx not in uniques["bonds"]:
+                        uniques["bonds"].append(bond_idx)
+
+        return uniques
+
+    def _draw_mapping(self):
+        mol1_uniques = self._get_unique_bonds_and_atoms(self.mol1_to_mol2)
+
+        # invert map
+        mol2_to_mol1_map = {v: k for k, v in self.mol1_to_mol2.items()}
+
+        mol2_uniques = self._get_unique_bonds_and_atoms(mol2_to_mol1_map)
+
+        atoms_list = [
+            mol1_uniques["atoms"] + mol1_uniques["elements"],
+            mol2_uniques["atoms"] + mol2_uniques["elements"],
+        ]
+
+        # highlight core element changes differently from unique atoms
+        red = (1, 0, 0, 0.5)
+        green = (0, 1, 0, 1)
+
+        at1_colours = {}
+        for at in mol1_uniques["elements"]:
+            at1_colours[at] = green
+
+        at2_colours = {}
+        for at in mol2_uniques["elements"]:
+            at2_colours[at] = green
+
+        atom_colours = [at1_colours, at2_colours]
+
+        bonds_list = [mol1_uniques["bonds"], mol2_uniques["bonds"]]
+
+        d2d = Chem.Draw.rdMolDraw2D.MolDraw2DCairo(600, 300, 300, 300)
```

Comment:
I think that this specific class can be replaced with others, e.g., `MolDraw2DSVG`, `MolDraw2DQt` -- you're just using the `MolDraw2D` API, right?

In that case, maybe allow the user to pass an `d2d` as an instance of a `MolDraw2D` to the method (default `None` creates what you have here). That allows this code to be reused for other duck-typed canvases.